File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/FhirPathCache.java
package org.opencds.cqf.fhir.utility;

import ca.uhn.fhir.context.FhirContext;
import ca.uhn.fhir.context.FhirVersionEnum;
import ca.uhn.fhir.fhirpath.IFhirPath;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class FhirPathCache {

    private FhirPathCache() {}

    private static final Map<FhirVersionEnum, IFhirPath> CACHE = new ConcurrentHashMap<>();

    public static IFhirPath cachedForContext(FhirContext fhirContext) {
        return CACHE.computeIfAbsent(fhirContext.getVersion().getVersion(), x -> fhirContext.newFhirPath());
    }

    public static IFhirPath cachedForVersion(FhirVersionEnum fhirVersionEnum) {
        return CACHE.computeIfAbsent(fhirVersionEnum, x -> x.newContext().newFhirPath());
        // when we upgrade to HAPI FHIR 5.0.0, we can use the following line
        //        return CACHE.computeIfAbsent(fhirVersionEnum, fhirVersion -> FhirContext.forVersion(fhirVersion)
        //                .newFhirPath());
    }
}

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/FhirPathCache.kt
package org.opencds.cqf.fhir.utility

import ca.uhn.fhir.context.FhirContext
import ca.uhn.fhir.context.FhirVersionEnum
import ca.uhn.fhir.fhirpath.IFhirPath
import java.util.concurrent.ConcurrentHashMap

object FhirPathCache {
    private val CACHE: MutableMap<FhirVersionEnum, IFhirPath> = ConcurrentHashMap()

    @JvmStatic
    fun cachedForContext(fhirContext: FhirContext): IFhirPath {
        return CACHE.computeIfAbsent(fhirContext.version.version) { x -> fhirContext.newFhirPath() }
    }

    @JvmStatic
    fun cachedForVersion(fhirVersionEnum: FhirVersionEnum): IFhirPath {
        return CACHE.computeIfAbsent(fhirVersionEnum) { x ->
            FhirContext.forCached(x).newFhirPath()
        }
    }
}
